Show metrics on pod details page

spadgett opened the pull request:
Adds memory usage and CPU usage charts to the pod details page. Requires hawkular-metrics, see

https://github.com/mwringe/origin-metrics

![screen shot 2015-10-12 at 5 11 11 pm](https://cloud.githubusercontent.com/assets/1167259/10438944/b8989292-7104-11e5-9b0f-02ccf7b1793d.png)

Follow on work:

- Consider other layouts (displaying a full spark chart beside the donut chart at wide screen widths?)
- Show best units for memory usage instead of always MiB
- Allow custom date ranges

/cc @jwforres @mwringe

jwforres commented:
@spadgett hmm debating whether uptime should be the first chart... 99.9% of the time (hopefully) that chart will be completely green.  But when its not completely green, that is pretty important...

Also the padding above / below the headers seems wrong, the header is too close to the previous chart and seems too far away from the chart it goes with.

jwforres commented:
Does the container dropdown become just text if there is only one item in the dropdown?   I personally hate dropdowns with one item in them, because I feel like it wastes my time when I open it and there arent any other options for me to choose

jwforres commented:
I think the time dropdown is probably good for now.  We might want to think about the angular ui timepicker or something later ( https://angular-ui.github.io/bootstrap/ )

Are you able to get a specific range, like from this start time and then ending at a specific time?  Just thinking about that zoom in feature I showed you the other day.  (not for this phase of impl of course)

spadgett commented:
I'll change the drop down to text if there's one container.

You can specify any time range. I don't think it would be too difficult to add a time picker, maybe show it if you click a link to allow any time range?

jwforres commented:
yeah i think that makes sense, an  "or choose a time..." link or something like that, so you have quick and easy options, or more advanced control if you want it.

spadgett commented:
I've added in a donut chart for memory limits (if set), and I've updated the CPU usage chart to use millicores. We could use CPU utilization %, but it might be confusing because it could be more than 100% if you have more than one core.

spadgett commented:
Is "Usage" a better tab title?

jwforres commented:
Monitoring? Or does that imply other stuff like alerting.
On Oct 9, 2015 4:56 PM, "Sam Padgett" <notifications@github.com> wrote:

> Is "Usage" a better tab title?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4999#issuecomment-146984144>.
>

spadgett commented:
Rebased to use #4954, which has merged.

jwforres commented on assets/app/scripts/directives/metrics.js:
```diff
@@ -0,0 +1,257 @@
+        var getMemoryLimit = $parse('resources.limits.memory');
+        var getCPULimit = $parse('resources.limits.cpu');
+
+        function bytesToMB(value) {
```
are we going to want to convert to GB in some cases?  Also I think its actually MiB not MB that the resource limit is stored as. At least that is the case on the limit range object.  @derekwaynecarr can you confirm what units memory is in on the pod/container.

spadgett replied:
Limit can be specified with different units. Right now I'm converting everything to MB (including the resource limit), but eventually I want to show the best unit based on the data.

spadgett replied:
Looks like it comes back with different units in the yaml as well.

```
➜  origin git:(ui-metrics) oc get limits limits -n project1 -o yaml
apiVersion: v1
kind: LimitRange
metadata:
  creationTimestamp: 2015-10-09T17:28:54Z
  name: limits
  namespace: project1
  resourceVersion: "52606"
  selfLink: /api/v1/namespaces/project1/limitranges/limits
  uid: 36bfa306-6eab-11e5-91f6-080027c5bfa9
spec:
  limits:
  - max:
      cpu: "2"
      memory: 1Gi
    min:
      cpu: 200m
      memory: 6Mi
    type: Pod
  - default:
      cpu: 300m
      memory: 200Mi
    defaultRequest:
      cpu: 200m
      memory: 100Mi
    max:
      cpu: "2"
      memory: 1Gi
    maxLimitRequestRatio:
      cpu: "10"
    min:
      cpu: 100m
      memory: 4Mi
    type: Container
```

spadgett replied:
https://docs.openshift.org/latest/dev_guide/compute_resources.html#compute-resource-units

spadgett replied:
I switched the memory units to display MiB for now at least.

derekwaynecarr replied:
@spadgett its good to know when a question can be answered with the documentation that I wrote ;-)

jwforres commented on assets/app/scripts/directives/metrics.js:
```diff
@@ -0,0 +1,257 @@
+          case 'cpu/usage':
+            var cpuLimit = getCPULimit(container);
+            if (cpuLimit) {
+              // Remove the 'm' unit at the end.
```
i thought the usageValueFilter already did that?

spadgett replied:
Yeah, this CPU limit code is not complete. Does usageValueFilter deal with millicores?

spadgett replied:
(By the way, what's here is not correct, just a temporary hack to get something working. I still need to handle units properly and have a TODO.)

jwforres replied:
It will strip the m off. It has to do that for limit ranges on the settings
page.
On Oct 12, 2015 9:26 AM, "Sam Padgett" <notifications@github.com> wrote:

> In assets/app/scripts/directives/metrics.js
> <https://github.com/openshift/origin/pull/4999#discussion_r41755497>:
>
> > +            sparklineConfig.axis.x.tick.format = '%a %H:%M';
> > +          }
> > +
> > +          var limit;
> > +          switch (metric.id) {
> > +          case 'memory/usage':
> > +            var memLimit = getMemoryLimit(container);
> > +            if (memLimit) {
> > +              // Convert to MB. usageValueFilter returns bytes.
> > +              limit = bytesToMB(usageValueFilter(memLimit));
> > +            }
> > +            break;
> > +          case 'cpu/usage':
> > +            var cpuLimit = getCPULimit(container);
> > +            if (cpuLimit) {
> > +              // Remove the 'm' unit at the end.
>
> Yeah, this CPU limit code is not complete. Does usageValueFilter deal with
> millicores?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4999/files#r41755497>.
>

spadgett replied:
> It will strip the m off. It has to do that for limit ranges on the settings page.

Yeah, just switched the CPU limit code to use the filter.

jwforres commented on assets/app/scripts/directives/metrics.js:
```diff
@@ -0,0 +1,257 @@
+            if (convertUnitsFn) {
+              converted = convertUnitsFn(lastDataPoint);
+            }
+            chartData.used = d3.round(converted || lastDataPoint, 1);
```
how often is this getting updated?  does this whole function run every time we get a new data point back?

spadgett replied:
`updateChart` runs once per metric every 30 seconds right now.

jwforres commented on assets/app/scripts/services/auth.js:
```diff
@@ -118,7 +118,7 @@ angular.module('openshiftConsole')
       requestRequiresAuth: function(config) {
         // TODO: replace with real implementation
-        var requiresAuth = config.url.toString().indexOf("api/") > 0;
+        var requiresAuth = (config.auth && config.auth.requiresAuth) || config.url.toString().indexOf("api/") > 0;
```
@liggitt auth.js change, review

spadgett replied:
@liggitt not ready for review yet, I still need to change data service.

spadgett replied:
@liggitt OK, the auth changes are in auth.js and data.js if you want to take a look. (The rest of the pull request is not ready for review yet.)

jwforres commented on assets/app/scripts/services/metrics.js:
```diff
@@ -0,0 +1,122 @@
+    var templateByMetric = {
+      "cpu/usage": COUNTER_TEMPLATE,
+      "memory/usage": GAUGE_TEMPLATE,
+      "memory/page_faults": RATE_TEMPLATE,
```
we aren't showing page faults though right, should we clean up what we arent using or do you think we would have these soon

spadgett replied:
Yes, I plan to remove this if it's not something we ever want to show.

jwforres commented on assets/app/scripts/services/metrics.js:
```diff
@@ -0,0 +1,122 @@
+
+      // Remove the first value since it can't be used CPU utilization.
+      // We want the same number of data points for all charts.
+      data.shift();
```
should we ask for the data point prior to the time range we asked for so that we dont drop the data point at start?

spadgett replied:
Yeah, I plan to refactor this still.

jwforres commented on assets/app/scripts/services/metrics.js:
```diff
@@ -0,0 +1,122 @@
+            metric: kwArgs.metric
+          }).toString();
+
+          Logger.log("Requesting metrics:", reqURL);
```
If this is going to generate a log every 30 seconds for each chart then we may want to make a sub logger like we do for auth that can be enabled independently from the main logger.

jwforres commented on assets/app/views/browse/pod.html:
```diff
@@ -80,11 +80,15 @@
-              <!--
-              <tab heading="Metrics" active="selectedTab.metrics">
-              Metrics content
+              <tab ng-if="metricsAvailable && pod.status.phase === 'Running'" heading="Metrics" active="selectedTab.metrics">
```
wouldn't we still have the historical metrics for recently stopped pods?  i might have a job that ran and I want to check its cpu utilization or something for the time that it was running.

spadgett replied:
I still need to verify what Hawkular keeps for terminated pods.

mwringe replied:
The data exists even after the pod is deleted. Metric data is stored with a time-to-live and its only then that the data will become unavailable.

liggitt commented on assets/app/scripts/services/auth.js:
```diff
@@ -117,8 +117,7 @@ angular.module('openshiftConsole')
-        // TODO: replace with real implementation
-        var requiresAuth = config.url.toString().indexOf("api/") > 0;
+        var requiresAuth = (config.auth && config.auth.requiresAuth);
         authLogger.log('AuthService.requestRequiresAuth()', config.url.toString());
```
log `requiresAuth` as well, not sure why I didn't before

liggitt commented on assets/app/scripts/services/data.js:
```diff
@@ -150,6 +150,9 @@ angular.module('openshiftConsole')
     this._getNamespace(resource, context, opts).then(function(ns){
       $http(angular.extend({
```
if the passed opts contains `opts.http.auth`, it'll overwrite our requiresAuth flag (extend doesn't merge recursively)

spadgett replied:
Should be fixed now by following you suggestion below

https://github.com/openshift/origin/pull/4999#discussion-diff-41781690

liggitt commented on assets/app/scripts/services/auth.js:
```diff
@@ -117,8 +117,7 @@ angular.module('openshiftConsole')
       requestRequiresAuth: function(config) {
-        // TODO: replace with real implementation
-        var requiresAuth = config.url.toString().indexOf("api/") > 0;
+        var requiresAuth = (config.auth && config.auth.requiresAuth);
```
what about just testing for the presence of an auth struct?

liggitt commented on assets/app/scripts/services/data.js:
```diff
@@ -150,6 +150,9 @@ angular.module('openshiftConsole')
     this._getNamespace(resource, context, opts).then(function(ns){
       $http(angular.extend({
         method: 'DELETE',
+        auth: {
```
if we do `auth:{}`, that's a signal auth is required
anyone who doesn't want auth would do `{http:{auth:null}}`
anyone who wants custom auth options could do `{http:{auth:{opt:value}}}` and it wouldn't matter that we overrode `{}` set by data.js

ajacobs21e commented:
Hey @spadgett  - It's late so here's my feedback dump :)

Is this using patternfly's donut chart? It looks bigger here. Maybe because it's on top of the graph and not beside it? Also theirs uses
        XX
of YY gb used
in the center of the donut which gives it more context. 

Is it more important to see how much is available than how much has been used?

Does memory always increase at a steady rate? If so a line graph isn't helpful.

Was it decided to not show multiple containers on one chart? 

Agreed on showing dropdown as text if only one option. 

Can you see this? https://issues.jboss.org/browse/HAWKULAR-590
See the images DatePicker.png & DatePicker2.png. I'd prefer one of those date picker styles - either a dropdown or buttons for preset relative times with a custom option to choose exact dates. Gabriel is doing usability testing on all of the options. We should consider using their design once it's finalized. I like what you have for now. 

Can you move the units for the Y axis to the left of the axis line? Having it overlap the chart is difficult to read. 

Thanks! Looking good.

spadgett commented:
@ajacobs21e Thanks as always for your feedback.

This is the patternfly utilization chart, which of course includes the donut chart.

http://angular-patternfly.rhcloud.com/#/api/patternfly.charts.directive:pfUtilizationChart

> Is it more important to see how much is available than how much has been used?

Fair question, I'm not sure. We do show the amount available above the chart. I personally find it confusing to fill a circle, but show a different number in the middle.

> Was it decided to not show multiple containers on one chart?

We had some concerns about scaling if there were more than a few containers. I'm not sure how the donut chart would look either with multiple containers.

> Gabriel is doing usability testing on all of the options. We should consider using their design once it's finalized. I like what you have for now.

I like having "Custom" in the dropdown. I'm not crazy about popovers in general, though... How do you change the values once you've set them if hidden in a popover?

Unfortunately we only have two days.

> Can you move the units for the Y axis to the left of the axis line? Having it overlap the chart is difficult to read.

I think so. I'll look at it.

What do you think about have the donut chart to the left and sparkline to the right for each metric?

ajacobs21e commented:
> This is the patternfly utilization chart, which of course includes the donut chart.
> http://angular-patternfly.rhcloud.com/#/api/patternfly.charts.directive:pfUtilizationChart

Ah ok, the angular site has more than the regular patternfly site. They're demoing stuff today so maybe they'll be including this. Thanks, I'll make sure to check the angular site from now on for the newest.

> Fair question, I'm not sure. We do show the amount available above the chart. I personally find it confusing to fill a circle, but show a different number in the middle.

The design now is more focused on amount used. The donut is visually so much stronger than the text below the header and then the donut doesn't say the percentage. My eyes first see the amount used within the donut, then see the donut is mostly full but not sure the exact percentage, and then after scanning around see the 80 available of 800. It makes more sense to me to either see 720 of 800 used or 90% of 800 used within the donut for more context. I wouldn't want to put "xx available" in the donut. 

> We had some concerns about scaling if there were more than a few containers. I'm not sure how the donut chart would look either with multiple containers.

Makes sense. Yeah we'd need a separate donut for each container so after ~3ish it would be too busy with a row of donuts.

>I like having "Custom" in the dropdown. I'm not crazy about popovers in general, though... How do you change the values once you've set them if hidden in a popover?

You'd have to navigate to custom again. We might be able to come up with something that wouldn't require two clicks to change the date but not in two days. 

> What do you think about have the donut chart to the left and sparkline to the right for each metric?

Yes that would look better.

spadgett commented:
@ajacobs21e The donut chart does let you hover to see a percentage. I'm not certain it lets me change the label in the middle. There might be an option I can pass to the directive as part of the c3.js config for a custom label. I'll look at it.

ajacobs21e commented:
Hmm yea the demo has used, available, or percentage (out of total). But not an option for used out of total. Do you think that would be the most useful to put in the donut?

liggitt commented on assets/app/config.js:
```diff
@@ -101,5 +101,5 @@ window.OPENSHIFT_CONFIG = {
   },
   loggingURL: "",
-  metricsURL: ""
+  metricsURL: "http://hawkular.127.0.0.1.xip.io:1080/hawkular/metrics"
```
remove from commit

spadgett replied:
> remove from commit

Yeah, I have a TODO for that :) Still a work in progress.

liggitt commented on assets/app/scripts/directives/metrics.js:
```diff
@@ -0,0 +1,255 @@
+        // Update immediately on display, then every 15 seconds or on options changes.
+        update();
+        intervalPromise = $interval(update, 15 * 1000, false);
+        scope.$watch('options', update, true);
```
I thought $watch could call the callback immediately for you

liggitt commented on assets/app/scripts/directives/metrics.js:
```diff
@@ -0,0 +1,255 @@
+
+        scope.$on('$destroy', function() {
+          if (intervalPromise) {
+            $interval.cancel(intervalPromise);
```
clear intervalPromise after canceling it

liggitt commented:
auth changes LGTM

spadgett commented:
@jwforres Please review.

spadgett commented:
Rebased.

[test]

spadgett commented:
Hm.

```
Synchronizing local sources
Synchronizing [origin@master] from origin...
Warning: Permanently added '54.175.194.201' (ECDSA) to the list of known hosts.
Synchronizing [source-to-image@master] from source-to-image...
fatal: Unable to create '/data/src/github.com/openshift/source-to-image-bare/refs/heads/master.lock': File exists.

If no other git process is currently running, this probably means a
git process crashed in this repository earlier. Make sure no other git
process is running and remove the file manually to continue.
fatal: The remote end hung up unexpectedly
fatal: The remote end hung up unexpectedly
Build step 'Execute shell' marked build as failure
```

[test]

spadgett commented:
Test error:

```
ERROR (0 secs / 0.176 secs)
PhantomJS 1.9.8 (Linux) DataService #url should generate a correct URL for {"resource":"pods","namespace":"foo","isWebsocket":true,"resourceVersion":"5"} FAILED
	Error: [$injector:modulerr] Failed to instantiate module openshiftConsole due to:
	Error: [$injector:modulerr] Failed to instantiate module patternfly.charts due to:
	Error: [$injector:nomod] Module 'patternfly.charts' is not available! You either misspelled the module name or forgot to load it. If registering a module ensure that you specify the dependencies as the second argument.
	http://errors.angularjs.org/1.3.8/$injector/nomod?p0=patternfly.charts
	    at /data/src/github.com/openshift/origin/assets/bower_components/angular/angular.js:1766
	    at ensure (/data/src/github.com/openshift/origin/assets/bower_components/angular/angular.js:1688)
	    at module (/data/src/github.com/openshift/origin/assets/bower_components/angular/angular.js:1984)
	    at /data/src/github.com/openshift/origin/assets/bower_components/angular/angular.js:4094
	    at forEach (/data/src/github.com/openshift/origin/assets/bower_components/angular/angular.js:323)
	    at loadModules (/data/src/github.com/openshift/origin/assets/bower_components/angular/angular.js:4120)
	    at /data/src/github.com/openshift/origin/assets/bower_components/angular/angular.js:4095
	    at forEach (/data/src/github.com/openshift/origin/assets/bower_components/angular/angular.js:323)
	    at loadModules (/data/src/github.com/openshift/origin/assets/bower_components/angular/angular.js:4120)
	    at createInjector (/data/src/github.com/openshift/origin/assets/bower_components/angular/angular.js:4004)
	    at workFn (/data/src/github.com/openshift/origin/assets/bower_components/angular-mocks/angular-mocks.js:2339)
	    at /data/src/github.com/openshift/origin/assets/bower_components/angular-mocks/angular-mocks.js:2322
	    at /data/src/github.com/openshift/origin/assets/test/spec/services/dataServiceSpec.js:9
```

spadgett commented:
Looks like I need to update karma.conf.js.

spadgett commented:
Fork AMI: https://ci.openshift.redhat.com/jenkins/job/fork_ami/101/

spadgett commented:
Trello card: https://trello.com/c/weyDN7S0

jwforres commented:
LGTM

spadgett commented:
Rebased to latest. @mwringe mentioned that the branch needs to current for the latest metrics setup to work. Building a new fork_ami.

spadgett commented:
Rebased.

jwforres commented:
[merge]
